### ISSUE TITLE
discussion on commits failing due to invalid line code

### ISSUE THREAD

**Issue by bmeneg** [zaquestion/lab#765]
When I comment in a commit with `lab mr discussion --commit <sha>` but in a line surrounding the actual code touched by the change I receive the following error:
```
2021/12/06 17:09:19 ERROR: note_common.go:108: POST https://gitlab.com/api/v4/projects/20363472/merge_requests/680/discussions: 400 {message: 400 Bad request - Note {:line_code=>["can't be blank", "must be a valid line code"]}}
```
For instance, the commit is: 
```diff
newfile: Makefile oldfile: Makefile
     @@ -1,3 +1,4 @@
   1 +# #test
1  2  # SPDX-License-Identifier: GPL-2.0
2  3  VERSION = 4
3  4  PATCHLEVEL = 18
```
If I add a comment after `VERSION = 4` the mentioned error happens.

@prarit is it expected? Is there a "valid scope" that the user is allowed to comment?

**Comment by prarit** [comment]
That's interesting.  So you attempted to comment on, for example, line 3?

**Comment by prarit** [comment]
Hmm ... I never tried it :) Let me try a few things.  I would have thought that should work given that line is technically part of the change.

**Comment by krobelus** [comment]
yeah this is some weirdness in the gitlab API - when commenting on context lines we need to pass two line numbers, the old and the new one. The fix below should work:

BTW I never realized that commenting on lines was already possible like this.

Sadly we have a misconfiguration at work that prevents us from using URL-encoded paths like `user%2frepo`, which is why I'm not using lab. In my scripts I have a workaround that uses the numeric project ID instead..

```diff
diff --git a/cmd/note_common.go b/cmd/note_common.go
index 9b64962..dc66b60 100644
--- a/cmd/note_common.go
+++ b/cmd/note_common.go
@@ -94,11 +94,9 @@ func noteRunFn(cmd *cobra.Command, args []string) {
 	createNote(rn, isMR, int(idNum), msgs, filename, linebreak, commit, true)
 }
 
-func createCommitNote(rn string, mrID int, sha string, newFile string, oldFile string, oldline int, newline int, comment string, block bool) {
-	linetype := "old"
+func createCommitNote(rn string, mrID int, sha string, newFile string, oldFile string, linetype string, oldline int, newline int, comment string, block bool) {
 	line := oldline
 	if oldline == -1 {
-		linetype = "new"
 		line = newline
 	}
 
@@ -149,11 +147,12 @@ func createCommitComments(project string, mrID int, commit string, body string,
 	oldfile := ""
 	oldLineNum := -1
 	newLineNum := -1
+	lineType := ""
 	comments := ""
 	for scanner.Scan() {
 		if strings.HasPrefix(scanner.Text(), "| newfile:") {
 			if comments != "" {
-				createCommitNote(project, mrID, commit, newfile, oldfile, oldLineNum, newLineNum, comments, block)
+				createCommitNote(project, mrID, commit, newfile, oldfile, lineType, oldLineNum, newLineNum, comments, block)
 				comments = ""
 			}
 			// read filename
@@ -166,34 +165,30 @@ func createCommitComments(project string, mrID int, commit string, body string,
 			}
 		} else if strings.HasPrefix(scanner.Text(), "|") {
 			if comments != "" {
-				createCommitNote(project, mrID, commit, newfile, oldfile, oldLineNum, newLineNum, comments, block)
+				createCommitNote(project, mrID, commit, newfile, oldfile, lineType, oldLineNum, newLineNum, comments, block)
 				comments = ""
 			}
 			// read line numbers
-			fs := strings.Split(scanner.Text(), " ")
-
-			oldLineNum = -1
-			newLineNum = -1
-			for _, f := range fs {
-				if f == "" || f == "|" || f == "@@" {
-					continue
+			fs := strings.Fields(scanner.Text())
+			if len(fs) < 3 {
+				continue
+			}
+			var err error
+			oldLineNum, err = strconv.Atoi(fs[1])
+			if err != nil {
+				continue
+			}
+			newLineNum, err = strconv.Atoi(fs[2])
+			if err != nil {
+				panic(err)
+			}
+			lineType = "context"
+			if len(fs) > 3 {
+				if strings.HasPrefix(fs[3], "-") {
+					lineType = "old"
 				}
-				val, err := strconv.Atoi(f)
-				if err != nil {
-					// NaN
-					if strings.HasPrefix(f, "+") {
-						newLineNum = oldLineNum
-						oldLineNum = -1
-					}
-					break
-				} else {
-					// Number
-					if oldLineNum == -1 {
-						oldLineNum = val
-					} else {
-						newLineNum = val
-						break
-					}
+				if strings.HasPrefix(fs[3], "+") {
+					lineType = "new"
 				}
 			}
 		} else {
diff --git a/internal/gitlab/gitlab.go b/internal/gitlab/gitlab.go
index 55f7483..13c2be0 100644
--- a/internal/gitlab/gitlab.go
+++ b/internal/gitlab/gitlab.go
@@ -1760,9 +1760,10 @@ func CreateMergeRequestCommitDiscussion(projID string, id int, sha string, newFi
 		PositionType: "text",
 	}
 
-	if linetype == "new" {
+	if linetype == "new" || linetype == "context" {
 		position.NewLine = line
-	} else {
+	}
+	if linetype == "old" || linetype == "context" {
 		position.OldLine = line
 	}
```

**Comment by bmeneg** [comment]
@krobelus don't you want to open a PR with this fix?
Otherwise I'll have to steal your solution :smile:

**Comment by krobelus** [comment]
Yeah please go ahead and steal it. I'll open a PR next time.

**Comment by prarit** [comment]
Needs a bit of massaging.  I get some panics and failures.  Looking at it now.

**Comment by prarit** [comment]
I looked at this and realized it was very similar to my first attempt at writing a parser for the diff.  This doesn't work :( because you can have a diff file like:

```

| newfile: Makefile oldfile: Makefile
|        @@ -5,6 +5,8 @@ SUBLEVEL = 0
|  5   5  EXTRAVERSION =
|  6   6  NAME = Merciless Moray
|  7   7  
|      8 +# space
|      9 +
|  8  10  #
|  9  11  # DRM backport version
| 10  12  #


```

```
+			fs := strings.Fields(scanner.Text())
+			if len(fs) < 3 {
+				continue
+			}
+			var err error
+			oldLineNum, err = strconv.Atoi(fs[1])
+			if err != nil {
+				continue
+			}
+			newLineNum, err = strconv.Atoi(fs[2])
+			if err != nil {
+				panic(err)
+			}


```

**Comment by krobelus** [comment]
> |  5   5  EXTRAVERSION =
> |  6   6  NAME = Merciless Moray

TBH this format looks a bit overly explicit. Why not use the standard diff format, and compute the various line numbers on the fly?
Feels like we are exposing implementation details that the user doesn't care about.

**Comment by prarit** [comment]
I want it to match the webUI output which shows those line numbers.  In any case I'm fixing your changeset now.  I think I have an easy way forward.

**Comment by bmeneg** [comment]
> I want it to match the webUI output which shows those line numbers. In any case I'm fixing your changeset now. I think I have an easy way forward.

Oh, I was doing that too :) will leave it to you.

> TBH this format looks a bit too explicit. Why not use the standard diff format, and compute the various line numbers on the fly? Feels like we are exposing implementation details that the user doesn't care about.

As @prarit said, it's not really implementation detail, but trying to match webui diff style. One problem with the usual diff format is identifying comments in between lines: it would require some education from users to enclose their comments with empty lines or use another special char to indicate that line isn't part of the actual diff, but a comment.. this type of situations. The `|` at the beginning of the line helps in this aspect, unless he decides to use this same char at the beginning of his comment, but then he's really trying to break things.

**Comment by bmeneg** [comment]
Fixed by #771
